### PR TITLE
crosscluster/logical: handle user-defined types in SQL mode

### LINKED ISSUE #132164: eval: create TypeResolver that can hydrate a Table with foreign type descriptors
When https://github.com/cockroachdb/cockroach/pull/131130 merges, LDR will use Import's type resolver to hydrate the source table descriptors with the source side type descriptors [here](https://github.com/cockroachdb/cockroach/blob/849d013c758c4b76cdd9c9e0372e71eab00acabe/pkg/ccl/crosscluster/logical/logical_replication_job.go#L363). While this works for the kv path in LDR, the sql ingestion path panics with the following message:
```
panic: invalid datum type given: USER DEFINED ENUM: my_enum, expected USER DEFINED ENUM: public.my_enum
```
This panic occurs because the source type and dest type OID's mismatch:
https://github.com/msbutler/cockroach/blob/butler-udt-check/pkg/sql/types/types.go#L2163

This oid's are compared because the sql path passes datums to the sql statement [here](https://github.com/msbutler/cockroach/blob/butler-udt-check/pkg/ccl/crosscluster/logical/lww_row_processor.go#L515)

To avoid this panic, we need to build a new kind of the type resolver which allows us to hydrate the source table descriptor with the source types, but that prevents this sort of panic. Perhaps this type resolver _also_ requires the dest type descriptors as input? I'm unsure.

Jira issue: CRDB-42859

Epic CRDB-41465

### PR DESCRIPTION
**Pull request by rafiss**
This PR also includes a cleanup refactor:

### importer,crosscluster: move type resolver to a different package

Since we're using this type resolver for more than just IMPORT, it makes
more sense to put it in a more fitting package.

----

### crosscluster/logical: handle user-defined types in SQL mode

Previously, using a user-defined type in SQL mode would fail because of
a low level check in the execution engine that verifies that the input
datum has the same type OID as the destination column.

This does not work for LDR, since the input datums come directly from a
source table in a different cluster, so user-defined types have
different OIDs.

This fixes it by setting the datum type to "unknown" before executing
the insert query. The "unknown" type is what is normally used when any
SQL statement is sent to CRDB without explicit type annotations/hints.
When the execution engine sees this type, it will perform an automatic
(and cheap) immutable assignment cast to change the datum to the
appropriate type.

fixes https://github.com/cockroachdb/cockroach/issues/132164
Release note (ops change): Logical replication streams that reference
tables with user-defined types can now be created with the `mode =
immediate` option.

### CODE REVIEW

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 1]
```diff
@@ -1936,7 +1936,7 @@ func TestUserDefinedTypes(t *testing.T) {
 	dbB.Exec(t, "INSERT INTO data (pk, val2) VALUES (2, (4, 'dog'))")
 
 	var jobAID jobspb.JobID
-	dbA.QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE data ON $1 INTO TABLE data", dbBURL.String()).Scan(&jobAID)
+	dbA.QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE data ON $1 INTO TABLE data WITH mode = validated", dbBURL.String()).Scan(&jobAID)
```
Comment (rafiss) [inline pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1939]: oops i meant to make this use both modes. fixing this now

File: pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go [hunk 2]
```diff
@@ -211,16 +210,9 @@ func createLogicalReplicationStreamPlanHook(
 
 		sourceTypes := make([]*descpb.TypeDescriptor, len(spec.TypeDescriptors))
 		for i, desc := range spec.TypeDescriptors {
-			// Until https://github.com/cockroachdb/cockroach/issues/132164 is resolved,
-			// we cannot allow user-defined types on the SQL ingestion path.
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:216]: i think this little block of code should be in the previous commit.

File: pkg/ccl/crosscluster/logical/lww_row_processor.go [hunk 3]
```diff
@@ -92,6 +92,9 @@ func (q *queryBuilder) AddRow(row cdcevent.Row) error {
 	}
 	if err := it.Datum(func(d tree.Datum, col cdcevent.ResultColumn) error {
 		if dEnum, ok := d.(*tree.DEnum); ok {
+			// Override the type to Unknown to avoid a mismatched type OID error
+			// during execution. Note that Unknown is the type used by default
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/lww_row_processor.go:96]: the diff in this file should be in the previous commit.

**Review by msbutler — APPROVED** [review]
Amazing! this fix is much lower impact than what we originally proposed.

File: pkg/ccl/crosscluster/logical/lww_row_processor.go (hunk 3 above)
Comment (rafiss) [reply pkg/ccl/crosscluster/logical/lww_row_processor.go:96]: done

File: pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go (hunk 2 above)
Comment (rafiss) [reply pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:216]: done

**Review by rafiss — COMMENTED** [review]
tftr!

bors r+

**Pull request merged**